caikit/core/model_management/local_model_trainer.py

```diff
@@ -0,0 +1,328 @@
+                    trained_model.save(self.save_path)
+            self._completion_time = datetime.now()
+            log.debug2("Completion time for %s: %s", self.id, self._completion_time)
+            if return_completion_time:
+                return self._completion_time
+            return trained_model
+
+    ## Interface ##
+
+    # Expression for parsing retention policy
+    _timedelta_expr = re.compile(
+        r"^((?P<days>\d+?)d)?((?P<hours>\d+?)h)?((?P<minutes>\d+?)m)?((?P<seconds>\d*\.?\d*?)s)?$"
+    )
+
+    def __init__(self, config: aconfig.Config, instance_name: str):
+        """Initialize with a shared dict of all trainings"""
+        self._instance_name = instance_name
+        self._use_subprocess = config.get("use_subprocess", False)
+        self._retention_duration = config.get("retention_duration")
+        if self._retention_duration is not None:
+            try:
+                log.debug2("Parsing retention duration: %s", self._retention_duration)
+                self._retention_duration = timedelta(
+                    **{
+                        key: float(val)
+                        for key, val in self._timedelta_expr.match(
+                            self._retention_duration
+                        )
+                        .groupdict()
+                        .items()
+                        if val is not None
+                    }
+                )
+            except AttributeError:
+                error(
+                    "<COR63897671E>",
+                    ValueError(
+                        f"Invalid retention_duration: {self._retention_duration}"
+                    ),
+                )
+
+        # The shared dict of futures and a lock to serialize mutations to it
+        self._futures = {}
+        self._futures_lock = threading.Lock()
+
+    def train(
+        self,
+        module_class: Type[ModuleBase],
+        *args,
+        save_path: Optional[str] = None,
+        save_with_id: bool = False,
+        external_training_id: Optional[str] = None,
+        **kwargs,
+    ) -> "LocalModelFuture":
+        """Start training the given module and return a future to the trained
+        model instance
+        """
+        # Always purge old futures
+        self._purge_old_futures()
+
+        # Create the new future
+        model_future = self.LocalModelFuture(
+            self._instance_name,
+            module_class,
+            *args,
+            save_path=save_path,
+            save_with_id=save_with_id,
+            external_training_id=external_training_id,
+            use_subprocess=self._use_subprocess,
+            **kwargs,
+        )
+
+        # Lock the global futures dict and add it to the dict
+        with self._futures_lock:
+            assert (
+                model_future.id not in self._futures
+            ), f"UUID collision for model future {model_future.id}"
```

Comment:
I smell a line that will never be unit test covered!

Reply:
That's why it's an `assert`! It _will_ be covered, but the branch will never be followed...

Reply:
> Only after generating 1 billion UUIDs every second for the next 100 years, the probability of creating just one duplicate would be about 50%. Or, to put it another way, the probability of one duplicate would be about 50% if every person on earth owned 600 million UUIDs.

I don't think we'll need this line, but I won't block the PR on it existing

Reply:
IIRC, the real reason I added this was the presence of `external_training_id`. We have no control over how those are generated, so it could be `lambda: "foo"`